server/routes/getValidatorDetail.ts

```diff
@@ -0,0 +1,83 @@
+/* eslint-disable guard-for-in */
+import { Request, Response, NextFunction } from 'express';
+import { factory, formatFilename } from '../../shared/logging';
+import { Op } from 'sequelize';
+function flatten(validators) {
+    validators = JSON.parse(JSON.stringify(validators));
+    const filteredValidators = validators.rows.map((row) => {
+        let dataObject = { ...row, ...row.Validator };
+        delete dataObject.Validator;
+        return dataObject;
+    });
+    return filteredValidators;
+}
+
+function formatData(searchCriteria) {
+    for (let key in searchCriteria) {
+        if (key === 'name' || key === 'stash_id') {
+            let temp = { ...searchCriteria };
+            delete searchCriteria[key];
+            searchCriteria = {
+                ...searchCriteria,
+                [key]: {
+
+                    [Op.iLike]: `%${temp[key]}%`
+                }
+            }
+        }
+    }
+    return searchCriteria;
+}
+export const getCurrentValidators = async (models, req: Request, res: Response, next: NextFunction) => {
+    let currentValidators: any = [];
+    let { pagination = { currentPageNo: 1, pageSize: 20 }, searchCriteria } = req.query;
+    searchCriteria = formatData(searchCriteria);
+    currentValidators = await models.HistoricalValidatorStats.findAndCountAll({
+        include: {
+            model: models.Validators
+        },
+        where: { ...searchCriteria, isElected: true,/*isLatest: true*/ },
+        offset: pagination?.pageSize * (pagination?.currentPageNo - 1),
+        limit: pagination?.pageSize,
+        order: [
+            ['stash_id', 'ASC']
+        ]
+    });
+    return res.json({
+        status: 'Success',
+        result: {
+            currentValidators: flatten(currentValidators),
+            pagination: {
+                ...pagination,
+                totalRecords: currentValidators?.count || 0
+            }
+        }
+    });
+};
+export const getWaitingValidators = async (models, req: Request, res: Response, next: NextFunction) => {
+    let waitingValidators: any = [];
+    let { pagination = { currentPageNo: 1, pageSize: 20 }, searchCriteria } = req.query;
+    searchCriteria = formatData(searchCriteria);
+    waitingValidators = await models.HistoricalValidatorStats.findAndCountAll({
```

Comment:
Getting the waiting validators should not come from the historical statistics table it should come from the validators table.

Reply:
Let me share the purpose of table I have understood.

**Validators table**: This table would contain all the validators currently present in the chain.

**Historical Validator Stats table:** This table would store all the relevant data for each validator (join with validator table),the old and updated history and for the routes .So the data(latest) visible on staking-ui screen would be fetched from this table including the GlobalStatistics and the tabs data i.e Current and Waiting validators.

**GlobalStats** would fetch all the latest data for all validators.

**Current/Waiting validators** would fetch data with client-side PAGINATION and SEARCH FUNCTIONALITY (SEARCH BY NAME OR ADDRESS) applicable.

Spec document :
https://www.notion.so/Move-to-backend-4d995500e32645e1aea7ff8a9ef6a920

Reply:
The spec doc looks blocked for my email drew@commonwealth.im now.